zerver/tests/test_message_flags.py

```diff
@@ -568,6 +569,48 @@ def test_track_active_mobile_push_notifications(
         mock_push_notifications.assert_called()
 
 
+class MarkAllAsReadEndpointTest(ZulipTestCase):
+    def test_mark_all_as_read_endpoint(self) -> None:
+        self.login("hamlet")
+        hamlet = self.example_user("hamlet")
+        othello = self.example_user("othello")
+        self.subscribe(hamlet, "Denmark")
+
+        for i in range(0, 4):
+            self.send_stream_message(othello, "Verona", "test")
+            self.send_personal_message(othello, hamlet, "test")
+
+        unread_count = (
+            UserMessage.objects.filter(user_profile=hamlet)
+            .extra(where=[UserMessage.where_unread()])
+            .count()
+        )
+        self.assertNotEqual(unread_count, 0)
+
+        # timeout() doesn't work in test environment with database operations
+        # and they don't get committed - so we need to replace it with a mock
+        # that just calls the function.
+        def mock_timeout(seconds: int, func: Callable[[], object]) -> object:
+            return func()
+
+        with mock.patch("zerver.views.message_flags.timeout", new=mock_timeout):
+            result = self.client_post("/json/mark_all_as_read", {})
```

Comment:
I don't like that I don't understand why this happens in tests. Is it just about doing db operations in a thread not working in tests for some reason? @alexmv  Do you have any idea? I can dig into this to figure it out if not

Reply:
The testsuite is running inside a transaction, so likely the code running on the other thread does successfully the change -- but it can't be seen in the test's transaction, due to transaction isolation.

The mock is probably helpful regardless, since we don't want to randomly have tests hit the timeout (due to the test machine being loaded) when we're not testing for the timeout path.  But it'd also be good to confirm my guess about the cause, so we can put a comment above the mock which doesn't lie.

I also note that this is the first use of the timeout function, _and threads in general_, in the serving path.  I think we'll want to exercise this carefully outside of tests, where the threading is in effect, to make sure it works as expected.